README
~/dev/py/vmmgr/README

2014-07-25

We are set up to operate in four zones: eu-west-1, us-east-1, us-west-1,
and us-west-2.  

A /20 is set up for each VPC.  This is 16 class Cs.  It must be properly
aligned.

A VPC has to be defined for each region used.  This is done by executing
a command like 
    ec2addvpc 192.168.240.0/20 --region us-west-2 >> vpcx

A security group has been defined for each region.  This is done by 
adding a line to make.groups after commenting out the preceding line(s)
and then executing the command.





CHANGES
# ~/dev/py/vmmgr/CHANGES

v0.5.22
    2018-03-19
        * correct corrupt .dvcz/builds                              SLOC 1889
v0.5.21
    2018-03-08
        * fix setup.py and tox config files                         SLOC 1888
v0.5.20
    2017-12-20
        * code cleanup; pylinting                                   SLOC 1891
v0.5.19
    2017-10-07
        * add projConfig.toml                                       SLOC 1916
v0.5.18
    2017-09-07
        * pep8; add tox.ini, *requirements.txt                      SLOC 1876
v0.5.17
    2017-07-29
        * rewrite setup.py, MANIFEST.in; minor edits cfg files      SLOC 1833
v0.5.16
    2017-06-24
        * restructure for pytest                                    SLOC 1828
v0.5.15
    2017-05-06
        * drop dist/* from repos; plus minor edits                  SLOC 1851
v0.5.14
    2017-01-28
        * simplify utilities using optionz:: dump_options()         SLOC 1689
v0.5.13
    2017-01-02
        * test_startup.py collects info on interfaces, blk devices  SLOC 1726
v0.5.12
    2017-01-01
        * botocore collects info on instances and volumes           SLOC 1741
v0.5.11
    2016-12-31
        * experiment with botocore                                  SLOC 1710
v0.5.10
    2016-12-30
        * more pylinting; minor bug fixes                           SLOC 1678
v0.5.9
    2016-10-08
        * pylint; elaborate Host API                                SLOC 1886
v0.5.8
    2016-10-07
        * add pylint, coverage                                      SLOC 1868
v0.5.7
    2016-08-28
        * fix license, setup.py                                     SLOC 1688
v0.5.6
    2016-06-29
        * add ghpDoc/{css,img}/                                     SLOC 1640
v0.5.5
    2016-06-02
        * add bkp2U, other bash scripts                             SLOC 1091
v0.5.4
    2016-05-04
        * added .dvcz, projCfg*                                     SLOC 1029
v0.5.3
    2016-03-28
        * add and execute auto8it                                   SLOC  964
v0.5.2
    2016-02-26
        * update LICENSE                                            SLOC  962
v0.5.1
    2016-02-01
        * minor edits                                               SLOC  962
v0.5.0
    2016-01-11
        * add -V/--showVersion, -z/--doNothing options              SLOC  552
        * skeletal vmKill, vmLanch, vmScrub, vmUpdate, vmVerify     SLOC  962
v0.4.9
    2016-01-10
        * vmInit, vmList converted to Python3
        * testSetup.py for python3, boto3 succeeds                  SLOC  543
v0.4.8
    2016-01-09
        * boto3 version of testSetup (skeletal)                     SLOC  543
v0.4.7
    2016-01-08
        * testSetup working again                                   SLOC  528
v0.4.6
    2016-01-03
        * begin conversion to boto3                                 SLOC  523
v0.4.5
    2015-12-26
        * fix dead links, edit TODO
v0.4.4
    2015-11-28
        * minor edits to *.md                                       SLOC  523
v0.4.3
    2015-09-16
        * corrected various control files, countem, MANIFEST.in     SLOC  523
        * largely rewrote docs                                      SLOC  523
v0.4.2
    2015-06-17
        * add installit, update .gitignore                          SLOC  429
v0.4.1
    2014-12-15
        * convert to Python 3                                       SLOC  389
v0.4.0
    2014-12-01
        * renamed GATEWAYS IGATEWAYS, gateways igateways            SLOC  389
v0.3.2
    2014-12-01
        * internal documentation for vmmgr/__init__.py              SLOC  389
    2014-11-30
        * rudimentary testSetup.py working                          SLOC  292
    2014-11-26
        * vmInit debugged
        * region list handling corrected in vmInit, vmList          SLOC  290
v0.3.1
    2014-11-25
        * correct typos, sketch out objectives                      SLOC  303
v0.3.0
    2014-10-12
        * skeletal vmList                                           SLOC  234

v0.2.0
    2014-09-10
        * introduced new classes: VMMgr, Host, EC2Host, LinuxBox,
            and functions to retrieve core parameters               SLOC  218
v0.1.1
    2014-08-22
        * minor corrections to tables in vmmgr/__init__.py
v0.1.0
    2014-07-16
        * initial import































test_setup.py
#!/usr/bin/env python3
# testSetup.py

""" Test the vm_mgr_setup() function. """

import sys
import unittest

import boto3
from botocore import session

from vmmgr import IGATEWAY_IDS, REGIONS, VPC_IDS, VPC_CIDRS


class TestSetup(unittest.TestCase):
    """ Test the vm_mgr_setup() function. """

    def setUp(self):
        pass

    def tearDown(self):
        pass

#   # boto3 version drops this
#   def old_test_vm_mgr_setup(self):
#       """ Python 2.7 version of the test code. """

#       for r_ndx, region in enumerate(REGIONS):
#           region_info = boto.ec2.get_region(region)
#           self.assertTrue(region_info is not None)

#           vpc_cnx = boto.vpc.VPCConnection(region=region_info)
#           self.assertTrue(vpc_cnx is not None)

#           print("%-9s %s" % (region_info, VPC_CIDRS[r_ndx]))

#           print("  gateway: %s" % IGATEWAY_IDS[r_ndx])

#           # "Resourcewarning: unclosed <ssl.SSLSocket ... ,443)>
#           #
#           igws = vpc_cnx.get_all_internet_gateways()
#           self.assertEqual(len(igws), 1)
#           self.assertEqual(str(igws[0]),
#                            'InternetGateway:' + IGATEWAY_IDS[r_ndx])

#           for z_ndx, zone in enumerate(ZONES[r_ndx]):
#               print(
#                   "    zone %s, cidr %s" %
#                   (zone, SUBNET_CIDRS[r_ndx][z_ndx]))
#               # sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
#               #        SUBNET_CIDRS[r_ndx][z_ndx],
#               #        availability_zone=zone)
#               # print "SUBNET: %z" % sub

#           # HACK, no apparent effect
#           vpc_cnx.close()
#           # END HACK

    # COMMENTED OUT, in effect
    def zzz_test_vm_mgr_setup_boto3(self):
        """ Python 3/boto3 version of the test code. """

        # ResourceWarning warning messages like
        #     "Resourcewarning: unclosed <ssl.SSLSocket ... ,443)>
        # are a known problem; there appears to be no way to easily
        # get rid of them.  Due to Python3/urllib3 connection pooling,
        # which is a Good Thing.

        # DEBUG
        print("BOTO3-BASED CODE")
        # END

        for r_ndx, r_name in enumerate(REGIONS):

            # DEBUG
            print(("\n%s" % r_name.upper()))
            # END
            # region_info = boto.ec2.get_region(r_name)
            # self.assertTrue(region_info is not None)

            ec2 = boto3.resource('ec2', region_name=r_name)

            # DEBUG
            for instance in ec2.instances.all():
                print("%s: tags '%s'" % (instance, instance.tags))
            continue                # XXX
            # END

            # DEBUG
            #print("region_info: %s" % region_info)
            #print("  cnx:      %s" % region_info.connection)
            #print("  endpoint: %s" % region_info.endpoint)
            # END

            # XXX THIS CODE IS NOW UNREACHABLE
            vpc = ec2.Vpc(VPC_IDS[r_ndx])    # gets a ResourceWarning
            self.assertTrue(vpc is not None)
            self.assertEqual(vpc.cidr_block, VPC_CIDRS[r_ndx])

            # alternative approach ----------------------------------

            # CLIENT --------------------------------------
            client = boto3.client('ec2', region_name=r_name)

            # GATEWAYS ------------------------------------
            desc = client.describe_internet_gateways()
            igws = desc['InternetGateways']
            self.assertEqual(len(igws), 1)
            igw = igws[0]
            # DEBUG
            print(("IGW: %s" % igws[0]['InternetGatewayId']))
            sys.stdout.flush()
            # END
            my_id = igw['InternetGatewayId']
            self.assertEqual(my_id, IGATEWAY_IDS[r_ndx])

            # AVAILABILITY ZONES --------------------------
            # for z_ndx, zone in enumerate(ZONES[r_ndx]):
            #    print("    zone %s, cidr %s" % (zone, SUBNET_CIDRS[r_ndx][z_ndx]))
            # SUBNETS -----------------------
            #    #sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
            #    #        SUBNET_CIDRS[r_ndx][z_ndx],
            #    #        availability_zone=zone)
            #    #print "SUBNET: %z" % sub

            # INSTANCES ---------------------

            # VOLUMES -----------------------

    def test_vm_mgr_setup_botocore(self):
        """ Python 3/botocore version of the test code. """

        # DEBUG
        print("\nBOTOCORE-BASED CODE")
        # END
        sss = session.get_session()

        for r_ndx, r_name in enumerate(REGIONS):

            # DEBUG
            print(("\n%s" % r_name.upper()))
            # END

            ec2 = sss.create_client('ec2', region_name=r_name)

            inst = ec2.describe_instances()['Reservations'][0]['Instances']
            icount = len(inst)
            print("There is/are %d instances" % icount)
            for ndx in range(icount):
                instance = inst[ndx]
                # there are 28 keys
                print('  instance ID:   %s' % instance['InstanceId'])
                print('  instance type: %s' % instance['InstanceType'])
                print('  image ID:      %s' % instance['ImageId'])
                print('  key name:      %s' % instance['KeyName'])
                print('  public IP:     %s' % instance['PublicIpAddress'])
                print('  private IP:    %s' % instance['PrivateIpAddress'])
                print('  subnet ID:     %s' % instance['SubnetId'])
                print('  root device:   %s' % instance['RootDeviceName'])

                ifaces = instance['NetworkInterfaces']
                print("  There is/are %d interfaces" % len(ifaces))
                for count, iface in enumerate(ifaces):
                    # there are 14 keys
                    print(
                        "    ID:            %s" %
                        iface['NetworkInterfaceId'])
                    print("    Status:        %s" % iface['Status'])
                    print("    SubnetId:      %s" % iface['SubnetId'])
                    print("    VpcId:         %s" % iface['VpcId'])
                    print("    priv ip addr:  %s" % iface['PrivateIpAddress'])
                    print("    association:")
                    assoc = iface['Association']
                    print("      IpOwnerId:   %s" % assoc['IpOwnerId'])
                    print("      PublicIp:    %s" % assoc['PublicIp'])
                    print("    groups:")
                    groups = iface['Groups']
                    for which, group in enumerate(groups):
                        print("      %d: id %s, name %s" % (
                            which, group['GroupId'], group['GroupName']))

                blkdev = instance['BlockDeviceMappings']
                print("  There is/are %d block devices" % len(blkdev))
                for count in range(len(blkdev)):
                    dev = blkdev[count]
                    name = dev['DeviceName']
                    ebs = dev['Ebs']
                    status = ebs['Status']
                    vol_id = ebs['VolumeId']
                    del_on_term = ebs['DeleteOnTermination']    # boolean
                    print("    %d name %-9s status %s vol_id %s" % (
                        count, name, status, vol_id))

            volumes = ec2.describe_volumes()['Volumes']
            vcount = len(volumes)
            print("  There are %d volumes" % vcount)
            for ndx in range(vcount):
                volume = volumes[ndx]
                # there are 9 keys
                print('  volume ID:   %s' % volume['VolumeId'])
                print('  volume type: %s' % volume['VolumeType'])
                print('  size:        %s' % volume['Size'])
                print('  state:       %s' % volume['State'])
                print('  zone:        %s' % volume['AvailabilityZone'])
                print('  attachments:')
                att = volume['Attachments'][0]
                print('    volume ID:   %s' % att['VolumeId'])
                print('    device:      %s' % att['Device'])
                print('    state:       %s' % att['State'])
                print('    instance ID: %s' % att['InstanceId'])
                print('    del on term  %s' % att['DeleteOnTermination'])

            continue                # XXX
            # END

            # DEBUG
            #print("region_info: %s" % region_info)
            #print("  cnx:      %s" % region_info.connection)
            #print("  endpoint: %s" % region_info.endpoint)
            # END

            # XXX THIS CODE IS NOW UNREACHABLE
            vpc = ec2.Vpc(VPC_IDS[r_ndx])    # gets a ResourceWarning
            self.assertTrue(vpc is not None)
            self.assertEqual(vpc.cidr_block, VPC_CIDRS[r_ndx])

            # alternative approach ----------------------------------

            # CLIENT --------------------------------------
            client = boto3.client('ec2', region_name=r_name)

            # GATEWAYS ------------------------------------
            desc = client.describe_internet_gateways()
            igws = desc['InternetGateways']
            self.assertEqual(len(igws), 1)
            igw = igws[0]
            # DEBUG
            print(("IGW: %s" % igws[0]['InternetGatewayId']))
            sys.stdout.flush()
            # END
            my_id = igw['InternetGatewayId']
            self.assertEqual(my_id, IGATEWAY_IDS[r_ndx])

            # AVAILABILITY ZONES --------------------------
            # for z_ndx, zone in enumerate(ZONES[r_ndx]):
            #    print("    zone %s, cidr %s" % (zone, SUBNET_CIDRS[r_ndx][z_ndx]))
            # SUBNETS -----------------------
            #    #sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
            #    #        SUBNET_CIDRS[r_ndx][z_ndx],
            #    #        availability_zone=zone)
            #    #print "SUBNET: %z" % sub

            # INSTANCES ---------------------

            # VOLUMES -----------------------

if __name__ == '__main__':
    # suppress warnings about unclosed sockets
    unittest.main(warnings='ignore')


tests/test_setup.py
#!/usr/bin/env python3
# testSetup.py

""" Test the vm_mgr_setup() function. """

import sys
import unittest

import boto3
from botocore import session

from vmmgr import IGATEWAY_IDS, REGIONS, VPC_IDS, VPC_CIDRS


class TestSetup(unittest.TestCase):
    """ Test the vm_mgr_setup() function. """

    def setUp(self):
        pass

    def tearDown(self):
        pass

#   # boto3 version drops this
#   def old_test_vm_mgr_setup(self):
#       """ Python 2.7 version of the test code. """

#       for r_ndx, region in enumerate(REGIONS):
#           region_info = boto.ec2.get_region(region)
#           self.assertTrue(region_info is not None)

#           vpc_cnx = boto.vpc.VPCConnection(region=region_info)
#           self.assertTrue(vpc_cnx is not None)

#           print("%-9s %s" % (region_info, VPC_CIDRS[r_ndx]))

#           print("  gateway: %s" % IGATEWAY_IDS[r_ndx])

#           # "Resourcewarning: unclosed <ssl.SSLSocket ... ,443)>
#           #
#           igws = vpc_cnx.get_all_internet_gateways()
#           self.assertEqual(len(igws), 1)
#           self.assertEqual(str(igws[0]),
#                            'InternetGateway:' + IGATEWAY_IDS[r_ndx])

#           for z_ndx, zone in enumerate(ZONES[r_ndx]):
#               print(
#                   "    zone %s, cidr %s" %
#                   (zone, SUBNET_CIDRS[r_ndx][z_ndx]))
#               # sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
#               #        SUBNET_CIDRS[r_ndx][z_ndx],
#               #        availability_zone=zone)
#               # print "SUBNET: %z" % sub

#           # HACK, no apparent effect
#           vpc_cnx.close()
#           # END HACK

    # COMMENTED OUT, in effect
    def zzz_test_vm_mgr_setup_boto3(self):
        """ Python 3/boto3 version of the test code. """

        # ResourceWarning warning messages like
        #     "Resourcewarning: unclosed <ssl.SSLSocket ... ,443)>
        # are a known problem; there appears to be no way to easily
        # get rid of them.  Due to Python3/urllib3 connection pooling,
        # which is a Good Thing.

        # DEBUG
        print("BOTO3-BASED CODE")
        # END

        for r_ndx, r_name in enumerate(REGIONS):

            # DEBUG
            print(("\n%s" % r_name.upper()))
            # END
            # region_info = boto.ec2.get_region(r_name)
            # self.assertTrue(region_info is not None)

            ec2 = boto3.resource('ec2', region_name=r_name)

            # DEBUG
            for instance in ec2.instances.all():
                print("%s: tags '%s'" % (instance, instance.tags))
            continue                # XXX
            # END

            # DEBUG
            # print("region_info: %s" % region_info)
            # print("  cnx:      %s" % region_info.connection)
            # print("  endpoint: %s" % region_info.endpoint)
            # END

            ##################################
            # XXX THIS CODE IS NOW UNREACHABLE
            ##################################

            vpc = ec2.Vpc(VPC_IDS[r_ndx])    # gets a ResourceWarning
            self.assertTrue(vpc is not None)
            self.assertEqual(vpc.cidr_block, VPC_CIDRS[r_ndx])

            # alternative approach ----------------------------------

            # CLIENT --------------------------------------
            client = boto3.client('ec2', region_name=r_name)

            # GATEWAYS ------------------------------------
            desc = client.describe_internet_gateways()
            igws = desc['InternetGateways']
            self.assertEqual(len(igws), 1)
            igw = igws[0]
            # DEBUG
            print(("IGW: %s" % igws[0]['InternetGatewayId']))
            sys.stdout.flush()
            # END
            my_id = igw['InternetGatewayId']
            self.assertEqual(my_id, IGATEWAY_IDS[r_ndx])

            # AVAILABILITY ZONES --------------------------
            # for z_ndx, zone in enumerate(ZONES[r_ndx]):
            #    print("    zone %s, cidr %s" % (
            #        zone, SUBNET_CIDRS[r_ndx][z_ndx]))
            # SUBNETS -----------------------
            #    #sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
            #    #        SUBNET_CIDRS[r_ndx][z_ndx],
            #    #        availability_zone=zone)
            #    # print "SUBNET: %z" % sub

            # INSTANCES ---------------------

            # VOLUMES -----------------------

    def test_vm_mgr_setup_botocore(self):
        """ Python 3/botocore version of the test code. """

        # DEBUG
        print("\nBOTOCORE-BASED CODE")
        # END
        sss = session.get_session()

        for r_ndx, r_name in enumerate(REGIONS):

            # DEBUG
            print(("\n%s" % r_name.upper()))
            # END

            ec2 = sss.create_client('ec2', region_name=r_name)

            inst = ec2.describe_instances()['Reservations'][0]['Instances']
            icount = len(inst)
            print("There is/are %d instances" % icount)
            for ndx in range(icount):
                instance = inst[ndx]
                # there are 28 keys
                print('  instance ID:   %s' % instance['InstanceId'])
                print('  instance type: %s' % instance['InstanceType'])
                print('  image ID:      %s' % instance['ImageId'])
                print('  key name:      %s' % instance['KeyName'])
                print('  public IP:     %s' % instance['PublicIpAddress'])
                print('  private IP:    %s' % instance['PrivateIpAddress'])
                print('  subnet ID:     %s' % instance['SubnetId'])
                print('  root device:   %s' % instance['RootDeviceName'])

                ifaces = instance['NetworkInterfaces']
                print("  There is/are %d interfaces" % len(ifaces))
                for iface in ifaces:
                    # there are 14 keys
                    print(
                        "    ID:            %s" %
                        iface['NetworkInterfaceId'])
                    print("    Status:        %s" % iface['Status'])
                    print("    SubnetId:      %s" % iface['SubnetId'])
                    print("    VpcId:         %s" % iface['VpcId'])
                    print("    priv ip addr:  %s" % iface['PrivateIpAddress'])
                    print("    association:")
                    assoc = iface['Association']
                    print("      IpOwnerId:   %s" % assoc['IpOwnerId'])
                    print("      PublicIp:    %s" % assoc['PublicIp'])
                    print("    groups:")
                    groups = iface['Groups']
                    for which, group in enumerate(groups):
                        print("      %d: id %s, name %s" % (
                            which, group['GroupId'], group['GroupName']))

                blkdev = instance['BlockDeviceMappings']
                print("  There is/are %d block devices" % len(blkdev))
                for ndx2, dev in enumerate(blkdev):
                    name = dev['DeviceName']
                    ebs = dev['Ebs']
                    status = ebs['Status']
                    vol_id = ebs['VolumeId']
                    # del_on_term = ebs['DeleteOnTermination']    # boolean
                    print("    %d name %-9s status %s vol_id %s" % (
                        ndx2, name, status, vol_id))

            volumes = ec2.describe_volumes()['Volumes']
            vcount = len(volumes)
            print("  There are %d volumes" % vcount)
            for ndx in range(vcount):
                volume = volumes[ndx]
                # there are 9 keys
                print('  volume ID:   %s' % volume['VolumeId'])
                print('  volume type: %s' % volume['VolumeType'])
                print('  size:        %s' % volume['Size'])
                print('  state:       %s' % volume['State'])
                print('  zone:        %s' % volume['AvailabilityZone'])
                print('  attachments:')
                att = volume['Attachments'][0]
                print('    volume ID:   %s' % att['VolumeId'])
                print('    device:      %s' % att['Device'])
                print('    state:       %s' % att['State'])
                print('    instance ID: %s' % att['InstanceId'])
                print('    del on term  %s' % att['DeleteOnTermination'])

            continue                # XXX
            # END

            # DEBUG
            # print("region_info: %s" % region_info)
            # print("  cnx:      %s" % region_info.connection)
            # print("  endpoint: %s" % region_info.endpoint)
            # END

            ##################################
            # XXX THIS CODE IS NOW UNREACHABLE
            ##################################

            vpc = ec2.Vpc(VPC_IDS[r_ndx])    # gets a ResourceWarning
            self.assertTrue(vpc is not None)
            self.assertEqual(vpc.cidr_block, VPC_CIDRS[r_ndx])

            # alternative approach ----------------------------------

            # CLIENT --------------------------------------
            client = boto3.client('ec2', region_name=r_name)

            # GATEWAYS ------------------------------------
            desc = client.describe_internet_gateways()
            igws = desc['InternetGateways']
            self.assertEqual(len(igws), 1)
            igw = igws[0]
            # DEBUG
            print(("IGW: %s" % igws[0]['InternetGatewayId']))
            sys.stdout.flush()
            # END
            my_id = igw['InternetGatewayId']
            self.assertEqual(my_id, IGATEWAY_IDS[r_ndx])

            # AVAILABILITY ZONES --------------------------
            # for z_ndx, zone in enumerate(ZONES[r_ndx]):
            #    print("    zone %s, cidr %s" % (
            #        zone, SUBNET_CIDRS[r_ndx][z_ndx]))
            # SUBNETS -----------------------
            #    #sub = vpc_cnx.create_subnet(VPC_IDS[r_ndx],
            #    #        SUBNET_CIDRS[r_ndx][z_ndx],
            #    #        availability_zone=zone)
            #    # print "SUBNET: %z" % sub

            # INSTANCES ---------------------

            # VOLUMES -----------------------


if __name__ == '__main__':
    # suppress warnings about unclosed sockets
    unittest.main(warnings='ignore')


src/vmmgr/__init__.py
# ~/py/vmmgr/__init__.py

""" Virtual Machine Manager (vmmgr) module. """

# import os
# import sys
# import time
# import boto3

__all__ = ['__version__', '__version_date__',
           'REGIONS', 'GROUP_IDS',
           'VPC_IDS', 'VPC_CIDRS',
           'ZONES',
           'SUBNET_IDS', 'SUBNET_CIDRS',
           'IGATEWAY_IDS',
           # FUNCTIONS -----------------------------------
           'valid_region',
           # CLASSES -------------------------------------
           'VMMgr', 'VMMgrError',
           'Host', 'EC2Host', 'LinuxBox', ]

__version__ = '0.5.22'
__version_date__ = '2018-03-19'

# CONSTANTS #########################################################
# Regions of interest at this time (at the end of 2016 there were 14 regions)
REGIONS = ['eu-west-1', 'us-east-1',
           'us-west-1', 'us-west-2', ]

# these are IDs for vpc-binders in each region
GROUP_IDS = ['sg-2c1ccd49', 'sg-56374833',
             'sg-f8a6579d', 'sg-adeb52c8', ]

VPC_IDS = ['vpc-e248bc87', 'vpc-e975af8c',
           'vpc-07e70762', 'vpc-6ffb050a', ]

VPC_CIDRS = ['192.168.192.0/20', '192.168.208.0/20',
             '192.168.224.0/20', '192.168.240.0/20', ]

# This is a small subset of zones actually available.
ZONES = [['eu-west-1a', ],
         ['us-east-1c', 'us-east-1d', ],
         ['us-west-1c', ],
         ['us-west-2c', ], ]

SUBNET_CIDRS = [['192.168.192.0/24', ],
                ['192.168.208.0/24', '192.168.209.0/24', ],
                ['192.168.224.0/24', ],
                ['192.168.240.0/24', ]]

SUBNET_IDS = [['subnet-4f9f6c38', ],
              ['subnet-9ef237e9', 'subnet-cd3ed994', ],
              ['subnet-8085b3c6', ],
              ['subnet-19e2eb5f', ], ]

IGATEWAY_IDS = ['igw-cf32d6aa', 'igw-fb3bff9e',
                'igw-948d91f6', 'igw-9809e2fd']

# a default route (and so igw) has been added to each table
ROUTE_TABLES = ['rtb-f1649994', 'rtb-5d17cc38',
                'rtb-6e4dae0b',
                # the main table, then the local
                ['rtb-cf7f84aa', 'rtb-d1708bb4'], ]

# This is not accurate (a) the 'main' associations are missing and
# (b) the eu-west 'main' association is in any case with the wrong
# rtb.  Returns a list rather than a single value; the list maps to
# zones rather than regions.
RTB_ASSOCS = [['rtbassoc-afc833ca', ],
              ['rtbassoc-de69c7bb', 'rtbassoc-9669c7f3', ],
              ['rtbassoc-e3769486', ],
              # this is for the local rtb
              ['rtbassoc-09d7176c', ], ]

# FUNCTIONS #########################################################


def valid_region(region):
    """
    Whether a region name is valid.

    :param region: The name of an EC2 region.
    :type region: str
    :return: True if parameter is the name of a valid EC2 region.
    :rtype: bool
    """
    valid = False
    for r__ in REGIONS:
        if region == r__:
            valid = True
            break
    return valid

# CLASS #############################################################


class VMMgrError(RuntimeError):
    pass


class VMMgr(object):
    """
    :ivar _regions: List of valid EC2 region names.
    :ivar _group_ids: A list of security group IDs (strings).
    :ivar _vpc_ids: A list of Virtual Private Cloud (VPC) IDs.
    :ivar _vpc_cidrS: List of CIDR block specs, dotted quad followed by prefix
        length.
    :ivar _zones: List of availability zones within region.
    :ivar _subnet_cidrs: List of CIDR blocks associated with respective subnets
    :ivar _subnet_ids: List of subnet IDs.
    :ivar _igateway_ids: List of internet gateways associated with respective
        subnets.
    """

    def __init__(self):
        self._regions = REGIONS
        self._group_ids = GROUP_IDS
        self._vpc_ids = VPC_IDS
        self._vpc_cidrs = VPC_CIDRS
        self._zones = ZONES
        self._subnet_cidrs = SUBNET_CIDRS
        self._subnet_ids = SUBNET_IDS
        self._igateway_ids = IGATEWAY_IDS
        self._route_tables = ROUTE_TABLES
        self._rtb_assocs = RTB_ASSOCS

    def region(self, ndx):
        """
        Region name.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: the EC2 region name corresponding to the index.
        :rtype: str
        """
        return self._regions[ndx]

    def group_id(self, ndx):
        """
        ID of virtual private cloud (VPD) for region.
        """

        return self._group_ids[ndx]

    def vpc_id(self, ndx):
        """
        ID of virtual private cloud (VPD) for region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: the virtual private cloud ID corresponding to the index.
        :rtype: str
        """
        return self._vpc_ids[ndx]

    def vpc_cidr(self, ndx):
        """
        Main CIDR block for the region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: the VPC's CIDR block (dotted quad/prefix length)
        :rtype: str
        """
        return self._vpc_cidrs[ndx]

    def zone(self, ndx):
        """
        Availability zones in use within the region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: zones in use within the region
        :rtype: list of str
        """
        return self._zones[ndx]

    def subnet_cidr(self, ndx):
        """
        CIDR blocks used by subnets within the rgion.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: CIDR blocks associated with respective subnets in region.
        :rtype: list of list of str
        """
        return self._subnet_cidrs[ndx]

    def subnet_id(self, ndx):
        """
        IDs associated with region subnets.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: IDs of subnets in region.
        :rtype: list of list of str
        """
        return self._subnet_ids[ndx]

    def igateway_id(self, ndx):
        """
        ID of internet gateway for a region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: ID of internet gateway for an EC2 region.
        :rtype: str
        """
        return self._igateway_ids[ndx]

    def route_table(self, ndx):
        """
        Route tables associated with a region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: list of route table IDs for subnets in the region.
        :rtype: list of str
        """
        return self._route_tables[ndx]

    def rtb_assoc(self, ndx):
        """
        Route table associations for a region.

        :param ndx: zero-based index of an EC2 region.
        :type ndx: int
        :raises IndexError: if there is no EC2 region with this index.
        :returns: list of route table association IDs for subnets in region.
        :rtype: list of str
        """
        return self._rtb_assocs[ndx]

# CLASS #############################################################


class Interface(object):

    def __init__(self, name, fqdn, ip_addr):
        self._name = name

        # need checks
        self._fqdn = fqdn
        # need checks
        self._ip_addr = ip_addr

    @property
    def name(self):
        """
        Return the interface's name (like eth0).  On a given host,
        interface names must be unique.
        """
        return self._name

    @property
    def fqdn(self):
        """
        Return the interface's fully qualified domain name (like example.com).

        On a given host FQDNs must be `unique.
        """
        return self._fqdn

    @property
    def ip_addr(self):
        """
        Return the interface's IP address (like 192.168.0.1).

        On a given host IP addresses must be unique.
        """
        return self._ip_addr


class Host(object):
    """
    Parent class for hosts of various kinds.

    Practice is to associate a pair (IP address, domain name) with
    each interface.  In anticipated use, hosts will have 1 to 3
    interfaces.  Interfaces have a name, but the names are not
    guaranteed to be stable.  In the real world, the same physical
    interface may have multiple IP addresses and domain names.
    Need to support IPv6 as well as IPv4.
    """

    def __init__(self, name, interfaces=None):

        self._name = name

        self._interfaces = []
        if interfaces:
            for iface in interfaces:
                self.add_interface(iface)

    @property
    def name(self):
        """
        Return the host's conventional name.

        This must be a valid identifier and must be unique within the
        network.  For now the first character of the name must be a
        letter (underscore OK) and otherwise alphanumeric.
        """
        return self._name

    def add_interface(self, iface):
        """
        Add an interface to the Host specification.

        This is a triple: name, fully qualified domain name, and IP address.
        Provisionally on a given host each of these is unique.  The FQDN and
        IP address must be unique within the network.
        """
        # XXX need better checks
        if iface is None:
            raise VMMgrError('attempt to add a null interface')
        if iface.name is None:
            raise VMMgrError('attempt to add a nameless interface')
        if iface.fqdn is None:
            raise VMMgrError('attempt to add an interface without FQDN')
        if iface.ip_addr is None:
            raise VMMgrError('attempt to add an interface without an IP addr')
        self._interfaces.append(iface)


class EC2Host(Host):
    """ Virtual machine on Amazon EC2 cloud as Host. """

    def __init__(self, fqdn):
        # pylint: disable=useless-super-delegation
        super().__init__(fqdn)


class LinuxBox(Host):
    """ Linux box (actual hardware on a subnet) as Host. """

    def __init__(self, fqdn):
        # pylint: disable=useless-super-delegation
        super().__init__(fqdn)
